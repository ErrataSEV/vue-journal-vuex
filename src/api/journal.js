
import axios from "axios";

const journalApi = axios.create({
    baseURL: 'https://vue-demos-36557-default-rtdb.firebaseio.com'
})

export default journalApi