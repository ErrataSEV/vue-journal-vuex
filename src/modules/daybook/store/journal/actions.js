// export default myAction = async ({ commit }) => {
//
// }

import journalApi from "@/api/journal"

export const loadEntries = async ({ commit }) => {

    const { data } = await journalApi.get('/entries.json')
    const entries = []

    if (!data) {
        commit('setEntries', [])
        return
    }

    for (const id of Object.keys(data)) {
        entries.push({
            id,
            ...data[id]
        })
    }
    
    // TODO: sort desc by date

    commit('setEntries', entries)
}

export const updateEntry = async ({ commit }, entry) => {
    const {id, ...body} = entry
    
    await journalApi.put(`/entries/${id}.json`, body)

    commit('updateEntry', { ...entry })
}

export const createEntry = async ({ commit }, entry) => {
    
    const { data } = await journalApi.post('/entries.json', entry)
    
    commit('addEntry', { id: data.name, ...entry })

    return data.name
}


export const deleteEntry = async ({ commit }, id) => {

    await journalApi.delete(`/entries/${id}.json`)

    commit('removeEntry', id)

}
