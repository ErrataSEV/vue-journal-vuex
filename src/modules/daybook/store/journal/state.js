export default () => ({
    isLoading: true,
    entries: [
        {
            id: new Date().getTime(),
            date: new Date().toDateString(),
            text: 'Occaecat nostrud elit et nostrud sit irure. Adipisicing quis dolore incididunt sunt aliqua laborum elit irure eiusmod. Sit elit tempor eu velit minim commodo elit. Magna laborum nostrud dolor ea nulla enim qui do duis qui sunt mollit.',
            picture: null,
        },
        {
            id: new Date().getTime()+100,
            date: new Date().toDateString(),
            text: 'Proident reprehenderit minim reprehenderit adipisicing eiusmod tempor minim velit Lorem eiusmod velit non. Excepteur minim voluptate et do ea. Velit mollit dolor esse enim ea cillum aute.',
            picture: null,
        },
        {
            id: new Date().getTime()+1000,
            date: new Date().toDateString(),
            text: 'Ea reprehenderit reprehenderit incididunt nostrud aute proident. Est cupidatat esse ea nostrud qui enim. Veniam et ad pariatur dolor mollit. Velit consectetur deserunt fugiat dolor ad ex enim ex irure sit ex non. Esse incididunt ad consequat eu officia veniam aliquip qui enim nulla enim duis. Do do mollit minim dolor consectetur sint ad est dolor magna aliqua aliqua. Culpa consequat cupidatat dolore id Lorem duis esse.',
            picture: null,
        }
    ]
})