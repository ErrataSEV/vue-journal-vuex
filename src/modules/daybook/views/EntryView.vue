<template>

    <template v-if="entry">
        <div class="entry-title d-flex justify-content-between p-2">
            <div>
                <span class="text-success fs-3 fw-bold">{{ day }}</span>
                <span class="mx-1 fs-3">{{ month }}</span>
                <span class="mx-2 fs-4 fw-light">{{ yearDay }}</span>
            </div>
            
            <div>
                <input type="file" @change="onSelectedImage" accept="image/*" ref="imageSelector" v-show="false">
                <button class="btn btn-danger mx-2" @click="onDeleteEntry" v-if="entry.id">
                    Delete
                    <i class="fa fa-trash-alt">
                    </i>
                </button>
                <button class="btn btn-primary" @click="onSelectImage">
                    Upload
                    <i class="fa fa-upload">
                    </i>
                </button>
            </div>
        </div>
        
        <hr>
        
        <div class="d-flex flex-column px-3 h-75">
            <textarea placeholder="What happened today?" v-model="entry.text"></textarea>
        </div>
        
        <img
            v-if="localImage"
            :src="localImage"
            alt="entry-picture"
            class="img-thumbnail"
        >
        <img v-if="entry.picture && !localImage" :src="entry.picture" alt="entry-picture" class="img-thumbnail">
    
    </template>

    <Fab icon="fa-save" @on:click="saveEntry"></Fab>
    
</template>

<script>
import { defineAsyncComponent } from 'vue';
import { mapGetters, mapActions } from 'vuex';
import Swal from 'sweetalert2';

import getDayMonthYear from "../helpers/getDayMonthYear";
import uploadImage from '@/modules/daybook/helpers/uploadImage';

export default {
    name: 'EntryView',
    props: {
        id: {
            type: String,
            required: true
        }
    },
    data() {
        return {
            entry: null,
            file: null,
            localImage: null
        }
    },
    components: {
        Fab: defineAsyncComponent(() => import('../components/FabComponent.vue'))
    },
    computed: {
        ...mapGetters('journal', ['getEntryById']),
        day() {
            const { day } = getDayMonthYear(this.entry.date)
            return day
        },
        month() {
            const { month } = getDayMonthYear(this.entry.date)
            return month
        },
        yearDay() {
            const { yearDay } = getDayMonthYear(this.entry.date)
            return yearDay
        },

    },
    methods: {
        ...mapActions('journal', ['updateEntry', 'createEntry', 'deleteEntry']),
        loadEntry() {
            let entry = {}
            if (this.id === 'new') {
                entry = {
                    text: '',
                    date: new Date().getTime()
                }
            } else {
                entry = this.getEntryById(this.id)
                if (!entry) return this.$router.push({ name: 'no-entry' })
            }

            this.entry = entry
        },
        async saveEntry() {

            Swal.fire({
                title: 'Waiting pls...',
                allowOutsideClick: false
            })
            Swal.showLoading()

            const picture = await uploadImage( this.file )

            this.entry.picture = picture

            if (this.entry.id) {
                await this.updateEntry(this.entry)
            } else {
                let newId = await this.createEntry(this.entry)
                this.$router.push({ name: 'entry', params: { id: newId } })
            }
            
            this.file = null
            Swal.fire('Saved!', 'Entry saved successfully')
        },
        async onDeleteEntry() {

            const { isConfirmed } = await Swal.fire({
                title: 'Are you sure to delete?',
                text: 'Once removed, you can\'t recovery',
                showDenyButton: true,
                confirmButtonText: 'Yes, I\'m sure'
            })

            if (isConfirmed) {
                Swal.fire({
                    title: 'Waiting pls...',
                    allowOutsideClick: false
                })
                Swal.showLoading()

                await this.deleteEntry(this.entry.id)
                this.$router.push({ name: 'no-entry' })

                Swal.fire('Deleted!', '', 'success')
            }

        },

        onSelectedImage(event) {
            const file = event.target.files[0];

            if (!file) {
                this.localImage = null
                this.file = null
                return
            }

            this.file = file
            const fr = new FileReader()
            fr.onload = () => this.localImage = fr.result
            fr.readAsDataURL( file )

        },
        onSelectImage() {
            this.$refs.imageSelector.click();
        }
    },
    created() {
        this.loadEntry()
    },
    watch: {
        id() {
            this.loadEntry()
        }
    }
}
</script>

<style lang="scss" scoped>
textarea {
    font-size: 20px;
    border: none;
    height: 100%;

    &:focus {
        outline: none;
    }
}

img {
    bottom: 150px;
    box-shadow: 0px 5px 10px rgba($color: #000000, $alpha: 0.2);
    height: 25%;
    position: fixed;
    right: 20px;
    width: 25%;
    widows: 200px;
}
</style>