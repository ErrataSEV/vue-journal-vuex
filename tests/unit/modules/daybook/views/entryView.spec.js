import { createStore } from "vuex"
import { shallowMount } from "@vue/test-utils";

import Swal from "sweetalert2";

import journal from '@/modules/daybook/store/journal';
import { journalState } from '../../../mock-data/test-journal-state';

import EntryView from '@/modules/daybook/views/EntryView';

const createVuexStore = ( initialState ) => createStore({
    modules: {
        journal: {
            ...journal,
            state: { ...initialState }
        }
    }
})

jest.mock('sweetalert2', () => ({
    fire: jest.fn(),
    showLoading: jest.fn(),
    close: jest.fn()
}))

describe('Test on EntryView', () => {

    const store = createVuexStore(journalState)
    store.dispatch = jest.fn()

    const mockRouter = {
        push: jest.fn()
    }
    let wrapper

    beforeEach(() => {
        jest.clearAllMocks()
        wrapper = shallowMount( EntryView, {
            props: {
                id: '-NRUTjCltIHNFHZB1z-h'
            },
            global: {
                mocks: {
                    $router: mockRouter
                },
                plugins: [ store ]
            }
        })
    })

    test('must redirct the user to new ', () => {
        const wrapper = shallowMount( EntryView, {
            props: {
                id: 'Este ID no existe en el STORE'
            },
            global: {
                mocks: {
                    $router: mockRouter
                },
                plugins: [ store ],
            }
        })

        expect( mockRouter.push ).toHaveBeenCalledWith({ name: 'no-entry' })
    })

    test('must show the correct entry', () => {
        expect( wrapper.html() ).toMatchSnapshot()
        expect( mockRouter.push ).not.toHaveBeenCalled()
    })

    test('must delete the entry and exit', async () => {

        Swal.fire.mockReturnValueOnce( Promise.resolve({ isConfirmed: true }) )

        await wrapper.find('.btn-danger').trigger('click')

        expect( Swal.fire ).toHaveBeenCalledWith({
            title: 'Are you sure to delete?',
            text: 'Once removed, you can\'t recovery',
            showDenyButton: true,
            confirmButtonText: 'Yes, I\'m sure'
        })

        expect( store.dispatch ).toHaveBeenCalledWith("journal/deleteEntry", "-NRUTjCltIHNFHZB1z-h")
        expect( mockRouter.push ).toHaveBeenCalled()
    })
})