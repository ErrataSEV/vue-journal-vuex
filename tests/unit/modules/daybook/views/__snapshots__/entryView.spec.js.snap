// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test on EntryView must show the correct entry 1`] = `
<div class="entry-title d-flex justify-content-between p-2">
  <div><span class="text-success fs-3 fw-bold">26</span><span class="mx-1 fs-3">Marzo</span><span class="mx-2 fs-4 fw-light">2023, Domingo</span></div>
  <div><input type="file" accept="image/*" style="display: none;"><button class="btn btn-danger mx-2"> Delete <i class="fa fa-trash-alt"></i></button><button class="btn btn-primary"> Upload <i class="fa fa-upload"></i></button></div>
</div>
<hr>
<div class="d-flex flex-column px-3 h-75"><textarea placeholder="What happened today?"></textarea></div>
<!--v-if-->
<img src="https://res.cloudinary.com/dsn8cnadp/image/upload/v1679864883/curso-vue/hfedsy8zfyp5lxulkk9a.jpg" alt="entry-picture" class="img-thumbnail">
<fab-stub icon="fa-save"></fab-stub>
`;
